lib/HLSL/DxilDebugInstrumentation.cpp

```diff
@@ -0,0 +1,723 @@
+      m_HandleForUAV,     // %dx.types.Handle, ; resource handle
+      m_CurrentIndex,            // i32 c0: index in bytes into UAV
+      Zero32Arg,        // i32 c1: unused
+      TheValue,
+      ZeroArg,        // unused values
+      ZeroArg,        // unused values
+      ZeroArg,        // unused values
+      WriteMask_X
+    });
+
+    m_RemainingReservedSpaceInBytes -= 4;
+    assert(m_RemainingReservedSpaceInBytes < 1024);  // check for underflow
+
+    if (m_RemainingReservedSpaceInBytes != 0) {
+      m_CurrentIndex = BC.Builder.CreateAdd(m_CurrentIndex, BC.HlslOP->GetU32Const(4));
+    }
+    else {
+        m_CurrentIndex = nullptr;
+    }
+  }
+}
+
+void DxilDebugInstrumentation::addInvocationStartMarker(BuilderContext & BC) {
+  DebugShaderModifierRecordHeader marker{ 0 };
+  reserveDebugEntrySpace(BC, sizeof(marker));
+
+  marker.Header.Details.SizeDwords = DebugShaderModifierRecordPayloadSizeDwords(sizeof(marker));;
+  marker.Header.Details.Flags = 0;
+  marker.Header.Details.Type = DebugShaderModifierRecordTypeInvocationStartMarker;
+  addDebugEntryValue(BC, BC.HlslOP->GetU32Const(marker.Header.u32Header));
+  addDebugEntryValue(BC, m_InvocationId);
+}
+
+template<typename ReturnType>
+void DxilDebugInstrumentation::addStepEntryForType(DebugShaderModifierRecordType RecordType, BuilderContext & BC, Instruction * Inst) {
+
+  DebugShaderModifierRecordDXILStep<ReturnType> step = {};
+  reserveDebugEntrySpace(BC, sizeof(step));
+
+  step.Header.Details.SizeDwords = DebugShaderModifierRecordPayloadSizeDwords(sizeof(step));
+  step.Header.Details.Type = static_cast<uint8_t>(RecordType);
+  addDebugEntryValue(BC, BC.HlslOP->GetU32Const(step.Header.u32Header));
+  addDebugEntryValue(BC, m_InvocationId);
+  addDebugEntryValue(BC, BC.HlslOP->GetU32Const(m_InstructionIndex++));
+
+  if (RecordType != DebugShaderModifierRecordTypeDXILStepVoid) {
+    addDebugEntryValue(BC, Inst);
+  }
+}
+
+void DxilDebugInstrumentation::addStepDebugEntry(BuilderContext & BC, Instruction * Inst) {
+  if (Inst->getOpcode() == Instruction::OtherOps::PHI) {
+    return;
+  }
+
+  Type::TypeID ID = Inst->getType()->getTypeID();
+
+  switch (ID) {
+  case Type::TypeID::StructTyID:
+  case Type::TypeID::VoidTyID:
+    addStepEntryForType<void>(DebugShaderModifierRecordTypeDXILStepVoid, BC, Inst);
+    break;
+  case Type::TypeID::FloatTyID:
+    addStepEntryForType<float>(DebugShaderModifierRecordTypeDXILStepFloat, BC, Inst);
+    break;
+  case Type::TypeID::IntegerTyID:
+    if (Inst->getType()->getIntegerBitWidth() == 64) {
+      addStepEntryForType<uint64_t>(DebugShaderModifierRecordTypeDXILStepUint64, BC, Inst);
+    }
+    else {
+      addStepEntryForType<uint32_t>(DebugShaderModifierRecordTypeDXILStepUint32, BC, Inst);
+    }
+    break;
+  case Type::TypeID::DoubleTyID:
+    addStepEntryForType<double>(DebugShaderModifierRecordTypeDXILStepDouble, BC, Inst);
+    break;
+  case Type::TypeID::FP128TyID:
+  case Type::TypeID::HalfTyID:
```

Comment:
HalfTyID will show up with true half support (and you may already see this with min16float as well).

Reply:
Thanks. I had intended to fill that one out and just plain forgot! Fixed.